### PR TITLE
misc(build): refactor viewer bundler into reusable GhPagesApp

### PR DESCRIPTION
**Pull request by connorjclark**
ref https://github.com/GoogleChrome/lighthouse/pull/11545/files#r503644206

* Provides interface for bundling an app for gh pages
* Uses simpler synchronous methods (instead of promisfying everything)
* removed service worker from viewer (doesn't do anything, no plan to add functionality, and I think it messed with my local dev every now and then...)

Next immediate use will be treemap app. If we ever port the scorecalc to this repo, it'd use this code too.

image

### CODE REVIEW

File: lighthouse-viewer/app/index.html [hunk 1]
```diff
@@ -45,13 +44,9 @@ <h1 class="viewer-placeholder__heading">Lighthouse Report Viewer</h1>
 
 <script src="https://www.gstatic.com/firebasejs/4.1.2/firebase-app.js"></script>
 <script src="https://www.gstatic.com/firebasejs/4.1.2/firebase-auth.js"></script>
-<script src="src/viewer.js"></script>
+<script src="src/bundled.js"></script>
 <script>
   window.addEventListener('DOMContentLoaded', main);
-
-  if ('serviceWorker' in navigator) {
-    navigator.serviceWorker.register('sw.js');
```
Comment (patrickhulce) [inline lighthouse-viewer/app/index.html:53]: we had a serviceworker? 😆

File: lighthouse-viewer/app/index.html [hunk 2]
```diff
@@ -10,10 +10,9 @@
   <meta charset="utf-8">
   <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">
   <title>Lighthouse Report Viewer</title>
-  <link rel="manifest" href="manifest.json">
```
Comment (patrickhulce) [inline lighthouse-viewer/app/index.html:13]: is this a difficult one to keep? seems harmless to copy the manifest over

File: package.json [hunk 3]
```diff
@@ -74,7 +74,8 @@
     "save-latest-run": "./lighthouse-core/scripts/save-latest-run.sh",
     "compile-proto": "protoc --python_out=./ ./proto/lighthouse-result.proto && mv ./proto/*_pb2.py ./proto/scripts || (echo \"❌ Install protobuf ≥ 3.7.1 to compile the proto file.\" && false)",
     "build-proto-roundtrip": "mkdir -p .tmp && cross-env PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION_VERSION=2 python proto/scripts/json_roundtrip_via_proto.py",
-    "static-server": "node lighthouse-cli/test/fixtures/static-server.js"
+    "static-server": "node lighthouse-cli/test/fixtures/static-server.js",
+    "serve-gh-pages": "cd dist/gh-pages && python -m SimpleHTTPServer"
```
Comment (patrickhulce) [inline package.json:78]: when should we be using `serve-gh-pages` vs the static-server?

**Review by patrickhulce — COMMENTED** [review]
this will be great for sourcemap app thanks :)

LGTM % few Qs

File: package.json (hunk 3 above)
Comment (connorjclark) [reply package.json:78]: static-server only does viewer b/c of a hard coded approval. Only for puppeteer test.

use serve-gh-pages if you want local viewer with a reasonable path (not prefixed with dist/) (or any app in gp-pages/)

Comment (patrickhulce) [reply package.json:78]: ah gotcha, throw this line into viewer's README then? (also can update the `dist/viewer` path reference while you're there 😃 )

**Review by patrickhulce — APPROVED** [review]
LGTM

**Comment by patrickhulce** [comment]
Vercel doesn't seem to like this though, and I can't pull up the details. Did something break there with our viewer deployment?

**Review by brendankenny — COMMENTED** [review]
Can you add some documentation here? It's a definite loss readability-wise otherwise. The current file is simple and more or less self contained. This new version drops any internal comments because the viewer-specific gotchas don't apply, but as a result it's less clear what's going on to build viewer and if you're writing a new "GhPagesApp", there are a bunch of implicit things going on. e.g. where do you put the main html file? (implicit vs the explicit css and js). Why are some js files included by default (and browserified) and not others? etc

**Comment by brendankenny** [comment]
(to answer the earlier [DRY question](https://github.com/GoogleChrome/lighthouse/pull/11545#discussion_r504137838), I'm a big fan of the rule of three, personally (master is currently at one gh-pages build file), and if Patrick wants to move to rollup he should just say so :P, but this is fine with some documentation so I don't have to read two build files and open devtools just to figure out what files are going to be included in the bundle :)

**Comment by patrickhulce** [comment]
> If we ever port the scorecalc to this repo, it'd use this code too.

and CPU throttling calc :)

> if Patrick wants to move to rollup he should just say so

Yes, for the record again, I think it would be better if we used any modern bundler that's maintained (rollup, webpack, parcel) instead of rolling our own in multiple places. Asking @connorjclark to pickup that effort for just treemap though seemed unreasonable. In my judgement I don't view this to be a loss to readability (it actually seemed to be an improvement IMO even if we have no additional copies at all. Viewer now calls something resembling a bundler API without needing to understand the implementation of a bundler just to make sense of the script). I'm sorry if everyone else on the team views this as wasted effort, but I do appreciate it greatly and think it's a win :)

**Comment by brendankenny** [comment]
> Viewer now calls something resembling a bundler API without needing to understand the implementation of a bundler just to make sense of the script

But that's the point I was making, you do still have to understand the implementation, just now it's in two different places and there are fewer comments :) Some scripts are manually added, some are pulled automatically, some are browserified, some are concatted, html just appears, etc.

**Comment by connorjclark** [comment]
Instead of automatically/magically adding (like adding `app/src/*.js` to the js), could instead do more explicit:

```
const app = new GhPagesApp({
    name: 'viewer',
    appDir: `${__dirname}/../lighthouse-viewer/app`,
    htmlReplacements: {
      '%%LIGHTHOUSE_TEMPLATES%%': htmlReportAssets.REPORT_TEMPLATES,
    },
    html: 'index.html', // ----------- NEW CODE ----------------
    javascripts: [
      await generatorJsPromise,
      htmlReportAssets.REPORT_JAVASCRIPT,
      fs.readFileSync(require.resolve('idb-keyval/dist/idb-keyval-min.js'), 'utf8'),
      'src/*', // ----------- NEW CODE ----------------
    ],
    stylesheets: [
      htmlReportAssets.REPORT_CSS,
      'styles/*', // ----------- NEW CODE ----------------
    ],
    assetPaths: [
      'images/**/*',
      'manifest.json',
    ],
  });
```

paths would use `appDir` as root. and be glob-able. Better?

**Comment by brendankenny** [comment]
nice, yeah, that does seem like a good change

**Comment by connorjclark** [comment]
cool, that works out a lot better.

File: build/gh-pages-app.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+*/`;
+
+/**
+ * @typedef BuildOptions
+ * @property {string} name
+ * @property {string} appDir
+ * @property {string} htmlPath
+ * @property {Array<{path: string} | string>} stylesheets
+ * @property {Array<{path: string} | string>} javascripts
+ * @property {string[]} assetPaths
```
Comment (brendankenny) [inline build/gh-pages-app.js:43]: it makes more sense to me to make all four of these `Array<{path: string} | string>` if some of them are going to go this way, but it's not a huge deal

File: build/gh-pages-app.js [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+* or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/`;
+
+/**
+ * @typedef BuildOptions
+ * @property {string} name
+ * @property {string} appDir
+ * @property {string} htmlPath
+ * @property {Array<{path: string} | string>} stylesheets
```
Comment (brendankenny) [inline build/gh-pages-app.js:41]: jsdocs for these (especially the ones with the options like this) please? :)

File: build/gh-pages-app.js (hunk 4 above)
Comment (connorjclark) [reply build/gh-pages-app.js:43]: Can't do it for assetPaths without having some way to interpret a literal string as an asset.  `{path: string}[]` might work?

Concating an array of html files doesn't make sense, but perhaps `{path: string} | string` could be allowed?

Comment (brendankenny) [reply build/gh-pages-app.js:43]: Oh right, cause assets are just copied over, nothing new is created.

Yeah, maybe having them be be `{path: string}` helps self document by making them consistent with the others, but it also makes sense how you have it here.

**Pull request merged**